### ISSUE TITLE
[bug] tasks are broken when using pnpm

### ISSUE THREAD

**Issue by gossi** [moonrepo/moon#195]
**Describe the bug**

I was giving moon a test-drive. It worked until I tried to execute a task (with eslint). 

**Steps to reproduce**

1. Clone `moonrepo/examples`
2. Switch from `npm` to `pnpm` in `.moon/workspace.yml`
3. In project root: `echo "strict-peer-dependencies=false" > .npmrc` (for sanity)
4. Create `pnpm-workspace.yaml` file with these contents:
	  
	```
	packages:
	  - "apps/*"
	  - "packages/*"
	```

5. Install `pnpm i`
6. Run `moon run mjs-package:lint`

**Expected behavior**

eslint will run

**Screenshots**

Here is the console output:

```
❯ moon run mjs-package:lint
▪▪▪▪ mjs-package:lint
eslint --ext .ts,.tsx,.cts,.mts,.js,.jsx,.cjs,.mjs --fix --report-unused-disable-directives --no-error-on-unmatched-pattern --exit-on-fatal-error --ignore-path ../../.eslintignore . (in ./packages/mjs-package)
/Users/gossi/code/playground/monorepo-examples/node_modules/.bin/eslint:2
basedir=$(dirname "$(echo "$0" | sed -e 's,\\,/,g')")
          ^^^^^^^

SyntaxError: missing ) after argument list
    at Object.compileFunction (node:vm:352:18)
    at wrapSafe (node:internal/modules/cjs/loader:1033:15)
    at Module._compile (node:internal/modules/cjs/loader:1069:27)
    at Module._extensions..js (node:internal/modules/cjs/loader:1159:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Module._load (node:internal/modules/cjs/loader:827:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:77:12)
    at node:internal/main/run_main_module:17:47

Node.js v18.0.0
▪▪▪▪ mjs-package:lint (297ms)

 ERROR  Process ~/.moon/tools/node/18.0.0/bin/node failed with a 1 exit code.

```

PS. This is a similar problem with other tools as well (not only `eslint` -  I also tried `ember-template-lint`, bcz of the project I had at hand).

**Environment**

Command `npx envinfo` failed to execute 😅 

I tried moon `v0.5` and `v0.6` - both failed.

**Comment by milesj** [comment]
Thanks, this will fix it once it lands https://github.com/moonrepo/moon/pull/199

**Comment by milesj** [comment]
This issue is now fixed.

There may still be issues with node module resolution in general, but I don't have enough knowledge about pnpm to know what's happening. 

If there are still pnpm issues, please create another issue.

**Comment by glamseam** [comment]
I tried using pnpm with the latest [moon/example](https://github.com/moonrepo/examples) and it did not work.

I tried
```
pnpm moon run vue-vite-app:dev
```

Error
```
thread 'tokio-runtime-worker' panicked at 'called `Option::unwrap()` on a `None` value', /Users/runner/work/moon/moon/crates/action/src/install_node_deps.rs:75:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

My environment
- macOS 12.4 (intel)
- pnpm 7.6.0

**Comment by milesj** [comment]
@glamseam Is this on moon v0.8? If not can you try updating.

**Comment by glamseam** [comment]
@milesj Yes, I am. I used moon v0.8.0.
I also get an error with `pnpm` but not with `npm`.

**Comment by milesj** [comment]
@glamseam Interesting, so it's failing here: https://github.com/moonrepo/moon/blob/master/crates/action/src/install_node_deps.rs#L75 

Which I'm then assuming fails here: https://github.com/moonrepo/moon/blob/master/crates/action/src/install_node_deps.rs#L23

What's your `workspace.yml` config file look like?

**Comment by glamseam** [comment]
@milesj My `workspace.yml`

```yaml
# Trigger CI: 2

$schema: 'https://moonrepo.dev/schemas/workspace.json'

projects:
  browser-package: 'packages/browser-package'
  cjs-package: 'packages/cjs-package'
  mjs-package: 'packages/mjs-package'
  nextjs-app: 'apps/nextjs-app'
  node-package: 'packages/node-package'
  remix-app: 'apps/remix-app'
  vue-vite-app: 'apps/vue-vite-app'

node:
  version: '18.5.0'
  packageManager: 'pnpm'
  addEnginesConstraint: true
  dedupeOnLockfileChange: true
  syncProjectWorkspaceDependencies: true
  syncVersionManagerConfig: 'nvm'

typescript:
  projectConfigFileName: 'tsconfig.json'
  rootConfigFileName: 'tsconfig.json'
  syncProjectReferences: true

vcs:
  manager: 'git'
  defaultBranch: 'master'

actionRunner:
  logRunningCommand: true

```
Only the `node.packageManager` has been changed.

```
- packageManager: 'npm'
+ packageManager: 'pnpm'
```

**Comment by milesj** [comment]
Thanks, I believe this will fix the problem: https://github.com/moonrepo/moon/pull/225

**Comment by glamseam** [comment]
@milesj Thank you.
I defined the version and tried it but still get the error.

```yaml
...
node:
  version: '18.5.0'
  packageManager: 'pnpm'
  pnpm:
    version: '7.6.0'
...
```
Error
```
▪▪▪▪ vue-vite-app:dev
vite dev (in ./apps/vue-vite-app)
[trace] moon:utils:process Running command ~/.moon/tools/node/18.5.0/bin/node --preserve-symlinks --title vue-vite-app:dev --unhandled-rejections throw ../../node_modules/dist/node/cli.js dev (in ~/moon-example/apps/vue-vite-app)
  MOON_CACHE_DIR=/Users/user-name/moon-example/.moon/cache
  MOON_PROJECT_ID=vue-vite-app
  MOON_PROJECT_ROOT=/Users/user-name/moon-example/apps/vue-vite-app
  MOON_PROJECT_RUNFILE=/Users/user-name/moon-example/.moon/cache/runs/vue-vite-app/runfile.json
  MOON_PROJECT_SOURCE=apps/vue-vite-app
  MOON_TARGET=vue-vite-app:dev
  MOON_TOOLCHAIN_DIR=/Users/user-name/.moon
  MOON_WORKING_DIR=/Users/user-name/moon-example
  MOON_WORKSPACE_ROOT=/Users/user-name/moon-example
  NODE_PATH=/Users/user-name/moon-example/node_modules/.pnpm/node_modules
node:internal/modules/cjs/loader:949
  throw err;
  ^

Error: Cannot find module '/Users/user-name/moon-example/node_modules/dist/node/cli.js'
    at Module._resolveFilename (node:internal/modules/cjs/loader:946:15)
    at Module._load (node:internal/modules/cjs/loader:787:27)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:77:12)
    at node:internal/main/run_main_module:17:47 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}

Node.js v18.5.0
▪▪▪▪ vue-vite-app:dev (103ms)
[trace] moon:action-runner:batch:3:1 Failed to run action RunTarget(vue-vite-app:dev) in 167.335862ms

 ERROR  Process ~/.moon/tools/node/18.5.0/bin/node failed with a 1 exit code.
```

Since moonrepo is running Node.js with the `--preserve-symlinks` flag, it adds `node-linker=hoisted` to the .npmrc.

My .npmrc
```
auto-install-peers=true
strict-peer-dependencies=false
node-linker=hoisted
```

[About pnpm node-linker](https://pnpm.io/npmrc#node-linker)

**Comment by milesj** [comment]
I'll re-open this.

pnpm seems to do things quite differently than npm/yarn. I'll look into testing all these edge cases as part of v0.9.

**Comment by milesj** [comment]
Tested this here: https://github.com/moonrepo/moon/pull/232

Was able to resolve issues with both linker types.

**Comment by milesj** [comment]
Released v0.9 that included these changes.

Thanks for being patient here. I've never used pnpm personally, so implementing accurate support has been... interesting.

**Comment by glamseam** [comment]
@milesj I have tried the new release and can confirm that it works fine with `pnpm`.
Thank you very much for your great work.

**Comment by milesj** [comment]
@glamseam Awesome, glad to hear! I'll close this for now. If any other issues arise, please file another.
